### PR TITLE
DEPS: require updated python-dateutil, openpyxl

### LINKED ISSUE #15184: DEPS: require modern openpyxl
we currently support as old as 1.6.1

we should bump this to 2.3 (released mid-2015).

then can remove the compat for openpyxl < 2 code and fully support styles.

### PR DESCRIPTION
**Pull request by jreback**
closes #15184 

reminder to change: https://github.com/conda-forge/pandas-feedstock on release

### CODE REVIEW

**Comment by jreback** [comment]
cc @neirbowj since you wrote most of the code I just ripped out :>

**Comment by neirbowj** [comment]
👍 It had a good run. I hope folks found it useful. I'm glad to see it retired.

And incidentally, no concerns for the FreeBSD port because the applicable dependencies are already at:

- `textproc/py-openpyxl` 2.4.9
- `deve/py-dateutil` 2.6.1

**Comment by jreback** [comment]
should we bump openpxl to even newer?

**Comment by neirbowj** [comment]
I'm probably not the right person to answer that question. I'm not an openpyxl user and haven't studied the codebase since the last time I got roped into responding to one of that project's minor updates that broke backwards compatibility. Therefore I cannot say whether a more aggressive version bump would be a good idea, bad idea, or indifferent.

From a port maintainer perspective, is there anything you'd like me to investigate?

**Comment by jreback** [comment]
i just want to make sure that i didn’t remove any relevant tests (that were not tied to earlier versions)

**Comment by neirbowj** [comment]
That would be non-trivial for me to verify, and I'm afraid can't justify spending time on it at the moment unless the test removal impacts the FreeBSD port.

**Comment by Themanwithoutaplan** [comment]
My suggestion would be to go with >= 2.4.0 as this includes utilities and styles for working with Pandas dataframes which should make integration code a lot easier and reliable. NB. Merged cells are **still** a problem but openpyxl 2.5 introduces something that, in time, will make these easier to work with.

2.4 also gives you the `ws.values` property which makes reading XLSX files a lot easier than the current xlrd spaghetti. xlrd is slightly faster on really big files due to a lower memory overhead but my suggestion would be to allow openpyxl in read-only mode (parallelisation of worksheets should be possible).

Would be happy to discuss this with anyone but haven't time to do it myself at the moment.

2.4 is stable and I haven't seen anything 2.3 specific issues for a good long time.

**Comment by jreback** [comment]
@Themanwithoutaplan thanks for the suggestion.

**Comment by Themanwithoutaplan** [comment]
http://openpyxl.readthedocs.io/en/default/pandas.html explains everything (based on the work we did together in Montreal).

**Comment by Themanwithoutaplan** [comment]
Also, support for a dedicated datetime celltype is coming 2.5. Removes some of the problems associated with Excel's moronic handling of dates and times.

**Comment by jreback** [comment]
@Themanwithoutaplan is this api supported in >= 2.4? (would you do a PR to change the pandas code)?

**Comment by Themanwithoutaplan** [comment]
New datetimes are 2.5 only. Not sure when I can do a PR. We are having a [sprint here in Düsseldorf](https://www.meetup.com/de-DE/Python-Meeting-Dusseldorf/events/243737124/) next week but I was hoping to be able work on something else…

I would like to get 2.5 out the door. It's just missing the code for adding/removing rows and columns.

**Comment by jreback** [comment]
any comments @jorisvandenbossche @TomAugspurger

**Review by jorisvandenbossche — COMMENTED** [review]
For openpyxl, it seems all code you removed was for versions < 2.2. Is there a reason to not just bump the minimum required to 2.2 instead of directly 2.4 in the PR?

For dateutil, do we win (code cleanup wise) a lot by not leaving the minimum version at eg 2.4 ? 

Further:

- Can you also update the io.rst docs? (there is a mention of the needed openpyxl version somewhere I think)

**Comment by jorisvandenbossche** [comment]
> For dateutil, do we win (code cleanup wise) a lot by not leaving the minimum version at eg 2.4 ?

Although, by the time 0.22 will be released, dateutil 2.5 will probably be almost 2 years old, so that is probably fine.

**Comment by jreback** [comment]
> For openpyxl, it seems all code you removed was for versions < 2.2. Is there a reason to not just bump the minimum required to 2.2 instead of directly 2.4 in the PR?

why? that's a pretty old version, I don't see a reason to not require newer things in developing libraries (esp there is a speedup coming), see the above comments.

**Comment by jorisvandenbossche** [comment]
openpyxl is just 1 year old, which is not that much. If it doesn't cost us anything to provide users the ability to use older version, I would do that (at the moment it is simply the check of the version that would need to be changed no?). If we would start using those newer features, then sure.

**Comment by Themanwithoutaplan** [comment]
Definitely drop support for openpyxl < 2.4. There were changes in styles in all 2.x releases. 2.4 was built to provide an API for Pandas so that the spaghetti code could be cleared out. < 2.4 is not maintained.

**Comment by jreback** [comment]
@jorisvandenbossche @TomAugspurger

File: doc/source/whatsnew/v0.22.0.txt [hunk 1]
```diff
@@ -54,8 +54,26 @@ Backwards incompatible API changes
...
+   | Package         | Minimum Version | Required |
+   +=================+=================+==========+
+   | python-dateutil | 2.5.0           |    X     |
+   +-----------------+-----------------+----------+
+   | pytz            | 2016.1          |    X     |
+   +-----------------+-----------------+----------+
+   | openpyxl        | 2.4.0           |          |
+   +-----------------+-----------------+----------+
+
+- :func:`Series.fillna` now raises a ``TypeError`` instead of a ``ValueError`` when passed a list, tuple or DataFrame as a ``value`` (:issue:`18293`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.22.0.txt]: merge issue here

**Comment by jorisvandenbossche** [comment]
Please also update io.rst (there are several mentions of openpyxl and its supported version)

What is the advantage of hard dropping support for older pytz versions? It does not cost us anything? (I mean in code complexity)

> Definitely drop support for openpyxl < 2.4. There were changes in styles in all 2.x releases. 2.4 was built to provide an API for Pandas so that the spaghetti code could be cleared out. < 2.4 is not maintained.

But we do not use this API at this point? As we only have special cased code for openpyxl < 2.2. If there have been many changes in style, for me that is a reason more to not drop support (as long as we don't use features of the newer releases), if people want to preserve the actual styling they have right now.

**Comment by jreback** [comment]
> What is the advantage of hard dropping support for older pytz versions? It does not cost us anything? (I mean in code complexity)

this is true, yeah just wanted to set some sort of minimum, though I guess its not really needed.

> But we do not use this API at this point? As we only have special cased code for openpyxl < 2.2. If there have been many changes in style, for me that is a reason more to not drop support (as long as we don't use features of the newer releases), if people want to preserve the actual styling they have right now.

not the point, there was too much churn in the < 2.4 versions.

**Comment by Themanwithoutaplan** [comment]
openpyxl used to have unreliable and incorrect style implementation so backwards-incompatible changes were essential. Not doing users any favours by keeping support for these around, especially as code coverage was never complete I've just managed to update my fork and hope to work on the > 2.4 implementation in the near future. Using the API will make it much easier in future to decide who's responsible.

**Comment by Themanwithoutaplan** [comment]
I'm not having a lot of fun getting tests to run: segfault with Python 3.6 on Mac OS and unable to run test_excel.py in isolation. I'd better check the docs.

**Comment by jreback** [comment]
@jorisvandenbossche

**Comment by Themanwithoutaplan** [comment]
I've started work on this.

**Comment by Themanwithoutaplan** [comment]
Looking at the implementation: is it going to be possible to work directly with a dataframe? openpyxl doesn't need the intermediate `ExcelCell` object.

**Comment by jreback** [comment]
@Themanwithoutaplan the implementation is orthogonal to us changing versions
what would hapoen if you had a version of openpyxl that can process things more efficiently 
then we could dispatch it that way, keeping existing code for older versions
if it passes all tests then that would be fine

so i assume this is not possible currently?

**Comment by Themanwithoutaplan** [comment]
I'm going to look at interfacing directly from `df.to_excel()` using `openpyxl.utils.dataframe.dataframe_to_rows()` This removes the need for an `ExcelWriter` class entirely. The older code could then be binned entirely.

**Comment by Themanwithoutaplan** [comment]
After some further investigation It looks to me that the best fit for the modern openpyxl API would be have a drop-in replacement for `ExcelFormatter` class.

**Pull request merged**